module/zfs/zvol.c

```diff
@@ -146,24 +140,33 @@ zvol_find_by_dev(dev_t dev)
 	return (NULL);
 }
 
-/*
- * Find a zvol_state_t given the name provided at zvol_alloc() time.
- */
 static zvol_state_t *
-zvol_find_by_name(const char *name)
+zvol_find_by_name_hash(const char *name, uint64_t hash)
 {
 	zvol_state_t *zv;
+	struct hlist_node *p;
 
 	ASSERT(MUTEX_HELD(&zvol_state_lock));
-	for (zv = list_head(&zvol_state_list); zv != NULL;
-	    zv = list_next(&zvol_state_list, zv)) {
-		if (strncmp(zv->zv_name, name, MAXNAMELEN) == 0)
+	hlist_for_each(p, ZVOL_HT_HEAD(hash)) {
+		zv = hlist_entry(p, zvol_state_t, zv_hlink);
+		if (zv->zv_hash == hash &&
+		    strncmp(zv->zv_name, name, MAXNAMELEN) == 0)
 			return (zv);
 	}
-
 	return (NULL);
 }
 
+/*
+ * Find a zvol_state_t given the name provided at zvol_alloc() time.
+ */
+static zvol_state_t *
+zvol_find_by_name(const char *name)
+{
+	uint64_t hash = zvol_name_hash(name);
```

Comment:
The local variable isn't really buying you anything here.  Why not make it a one liner.